src/cluster_import.c

```diff
@@ -0,0 +1,1915 @@
+                  (unsigned long long)c->id);
+        freeClientAsync(c);
+        return;
+    }
+    slotMigrationLink *link = (slotMigrationLink *)c->slot_migration_link;
+    if (!isSlotMigrationLinkInProgress(link)) {
+        return;
+    }
+    if (link->state != SLOT_IMPORT_FAILOVER_REQUESTED) {
+        serverLog(LL_WARNING,
+                  "Received CLUSTER SYNCSLOTS FAILOVER-GRANTED from link %.40s, "
+                  "but we never sent a failover request. Closing link.",
+                  link->linkname);
+        finishSlotMigrationLink(link, SLOT_MIGRATION_LINK_FAILED,
+                                "Unexpected state machine transition");
+        return;
+    }
+    updateSlotMigrationLinkState(link, SLOT_IMPORT_FAILOVER_GRANTED);
+    clusterDoBeforeSleep(CLUSTER_TODO_HANDLE_SLOT_MIGRATION);
+}
+
+/* Sent by the source to the target to deny final authorization for
+ * failover. */
+void clusterCommandSyncSlotsFailoverDenied(client *c) {
+    if (!c->flag.slot_import_source || !c->slot_migration_link) {
+        serverLog(LL_WARNING,
+                  "Received CLUSTER SYNCSLOTS FAILOVER-DENIED from client %llu, "
+                  "but the client is not a slot import source. Closing link.",
+                  (unsigned long long)c->id);
+        freeClientAsync(c);
+        return;
+    }
+    slotMigrationLink *link = (slotMigrationLink *)c->slot_migration_link;
+    if (!isSlotMigrationLinkInProgress(link)) {
+        return;
+    }
+    if (link->state != SLOT_IMPORT_FAILOVER_REQUESTED) {
+        serverLog(LL_WARNING,
+                  "Received CLUSTER SYNCSLOTS FAILOVER-DENIED from link %.40s, "
+                  "but we never sent a failover request. Closing link.",
+                  link->linkname);
+        finishSlotMigrationLink(link, SLOT_MIGRATION_LINK_FAILED,
+                                "Unexpected state machine transition");
+        return;
+    }
+    if (link->one_shot) {
+        serverLog(LL_WARNING,
+                  "Slot import link %.40s had failover denied from node %.40s "
+                  "(owner of slots [%s]). Failing import request.",
+                  link->linkname,
+                  link->nodename,
+                  link->slot_ranges_str);
+        finishSlotMigrationLink(link, SLOT_MIGRATION_LINK_FAILED, "Failover denied");
+        return;
+    }
+    serverLog(LL_WARNING,
+              "Slot import link %.40s had failover denied from node %.40s "
+              "(owner of slots [%s]). Going back to streaming incremental updates.",
+              link->linkname,
+              link->nodename,
+              link->slot_ranges_str);
+    updateSlotMigrationLinkStatusMessage(link, "Previous failover attempt denied from source");
+    updateSlotMigrationLinkState(link, SLOT_IMPORT_RECEIVE_STREAM);
+}
+
+/* Sent by the source to the target to notify of some unrecoverable error. */
+void clusterCommandSyncSlotsFail(client *c) {
```

Comment:
It seems like there is some left over terminology for slot sync, which I assume is what tencent called their implementation? Ideally we should converge on internal and external naming.